List editable choice

lutangar opened the pull request:
Add x-editable support for (non multiple) choice in lists, this feature was once mentioned in #1867 by @EmmanuelVella  and also requested in #2269.
Multiple choice may require more work though.

First thing was to move the `x-editable` types mapping to the **parameters**, *imho* its was pretty annoying to have it hard-coded in the twig extension. Especially if you need to add more editable types in your project.

The behaviour is quite similar to the way `booleans` were handled. 
```php
    protected function configureListFields(ListMapper $list)
    {
        $list
            ->add('status', 'choice',
                [
                    'editable' => true,
                    'choices' => ['Status1' => 'Alias1', 'Status2' => 'Alias2'],
                    'catalogue' =>'admin'
                ]
            );
    }
```

The tricky part is to transform the provided **choices** options into something that can be fed to the `x-editable` widget : 
### Choices
```php
['Status1' => 'Alias1', 'Status2' => 'Alias2']
```
### X-editable choices
```php
[['value' => 'Status1', 'text' => 'Alias1'], ['value' => 'Status2', 'text' => 'Alias2']]
```
This is working pretty well for `enum` like `array`, but It might be interesting to handle more inputs, like `ArrayCollection`. 
How would you guys do to handle such cases?

Any feedbacks welcome.

ConneXNL commented:
Yay!

webdevilopers commented:
:+1:

Soullivaneuh commented on Resources/config/twig.xml:
```diff
@@ -4,11 +4,34 @@
+
     <services>
-        <service id="sonata.admin.twig.extension" class="Sonata\AdminBundle\Twig\Extension\SonataAdminExtension">
+        <service id="sonata.admin.twig.extension" class="%sonata.admin.twig.extension.class%">
```
Do not use class parameter. This is not recommended anymore.

Soullivaneuh commented:
Could you please rebase your PR?

lutangar commented:
It's been a while, anyway the rebase is done and I took into account your notice.

Soullivaneuh commented:
Test are note passing.

Please fix CS: https://travis-ci.org/sonata-project/SonataAdminBundle/jobs/67817285#L209

lutangar commented:
Done.

Soullivaneuh commented:
PR need to be rebased again.

lutangar commented:
Done.

OskarStark commented:
Thank you

Am Freitag, 28. August 2015 schrieb Johan Dufour :

> Done.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sonata-project/SonataAdminBundle/pull/2484#issuecomment-135722834>
> .
>

OskarStark commented:
@lutangar can you please check the failing travis build and provide a fix?

@Soullivaneuh i'm not sure if this is related to this PR, can you please have a look, too?

lutangar commented:
@OskarStark @Soullivaneuh alright all checks have passed

OskarStark commented:
thank you!!

Soullivaneuh commented:
@rande What do you think of this feature?

OskarStark commented:
@lutangar 

in your initial post:
```[['value' => 'Status1', 'text': 'Alias1'], ['value' => 'Status2', 'text': 'Alias2']]```

i think this is wrong, isn't it?
```'text': 'Alias...```  i think the ```:``` should be ```=>```, right?

OskarStark commented:
you said, that this is for ```non-multiple``` choices, if we set ```mutliple=true``` what would happen?

i recommend the to ignore the x-editable, and you?

is this case tested?

OskarStark commented on Resources/views/CRUD/list_choice.html.twig:
```diff
@@ -11,6 +11,18 @@ file that was distributed with this source code.
+
 {% block field %}
 {% spaceless %}
     {% if field_description.options.choices  is defined %}
```
2 spaces between ```choices``` and ```id```

OskarStark commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -75,6 +75,11 @@ class SonataAdminExtensionTest extends \PHPUnit_Framework_TestCase
+    /**
+     * @var array
+     */
+    protected $xEditableTypeMapping;
```
```protected``` or ```private```?

OskarStark commented on Twig/Extension/SonataAdminExtension.php:
```diff
@@ -314,29 +320,53 @@ public function getUrlsafeIdentifier($model, AdminInterface $admin = null)
+    }
+
+    /**
+     * Return x-editable choices based on the field description choices options & catalogue options
```
```x-editable``` -> ```xEditable``` ?

```choices options``` -> ```choice options``` ?

OskarStark commented on Twig/Extension/SonataAdminExtension.php:
```diff
@@ -314,29 +320,53 @@ public function getUrlsafeIdentifier($model, AdminInterface $admin = null)
+    /**
+     * Return x-editable choices based on the field description choices options & catalogue options
+     * With the following choices option: ['Status1' => 'Alias1', 'Status2' => 'Alias2']
+     * The method will return : [['value' => 'Status1', 'text': 'Alias1'], ['value' => 'Status2', 'text': 'Alias2']].
```
```'text': '``` ?

core23 commented:
Ping @lutangar

lutangar commented:
@OskarStark I fixed the minor things and rebased the feature.

> you said, that this is for non-multiple choices, if we set mutliple=true what would happen?
> i recommend to ignore the x-editable option, and you?
> is this case tested?

In the multiple choices case, I suppose the widget would default to non multiple choices, but this is not tested.
I am not sure that the x-editable option should be ignored, the `getXEditableType` should remain stupid. Of course, ideally the full feature should cover all choices type and we wouldn't have to care about this... but that's definitely more work than I could afford at the moment.

OskarStark commented:
thank you @lutangar could you please add a screenshot how it looks like?
this would be awesome

@core23 what do you think?

core23 commented:
LGTM

core23 commented:
Please reabse @lutangar

lutangar commented:
@core23 rebase done.

core23 commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -76,6 +76,11 @@ class SonataAdminExtensionTest extends \PHPUnit_Framework_TestCase
     private $logger;
 
+    /**
+     * @var array
```
Could be ``string[]``

core23 commented on Twig/Extension/SonataAdminExtension.php:
```diff
@@ -36,6 +36,11 @@ class SonataAdminExtension extends \Twig_Extension implements \Twig_Extension_In
     /**
+     * @var array
+     */
+    private $xEditableTypeMapping = array();
```
Could be ``string[]``

core23 commented on Twig/Extension/SonataAdminExtension.php:
```diff
@@ -384,29 +393,53 @@ public function getUrlsafeIdentifier($model, AdminInterface $admin = null)
     }
 
     /**
+     * @param array $xEditableTypeMapping
```
Could be ``string[]``

greg0ire commented on Resources/views/CRUD/list_choice.html.twig:
```diff
@@ -11,9 +11,21 @@ file that was distributed with this source code.
 
 {% extends admin.getTemplate('base_list_field') %}
 
+{% set isEditable = field_description.options.editable is defined and field_description.options.editable and admin.isGranted('EDIT', object) %}
```
Please linebreak this line.

greg0ire commented on Twig/Extension/SonataAdminExtension.php:
```diff
@@ -405,4 +409,37 @@ public function getXEditableType($type)
+                foreach ($choices as $value => $text) {
+                    $xEditableChoices[] = array(
+                        'value' => $value,
+                        'text'  => $catalogue ? $fieldDescription->getAdmin()->trans($text, array(), $catalogue) : $text,
```
Please linebreak this line

greg0ire commented on Twig/Extension/SonataAdminExtension.php:
```diff
@@ -405,4 +409,37 @@ public function getXEditableType($type)
+    /**
+     * Return xEditable choices based on the field description choices options & catalogue options.
+     * With the following choice options: ['Status1' => 'Alias1', 'Status2' => 'Alias2']
+     * The method will return: [['value' => 'Status1', 'text' => 'Alias1'], ['value' => 'Status2', 'text' => 'Alias2']].
```
Please linebreak this line.

greg0ire commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -718,6 +746,132 @@ class="x-editable"
+        Status1
+    </span>
+</td>
+EOT
```
Nice job :+1:

greg0ire commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -718,6 +746,132 @@ class="x-editable"
+        data-value="Status1"
+        data-title="Data"
+        data-pk="12345"
+        data-url="/core/set-object-field-value?context=list&amp;field=fd_name&amp;objectId=12345&amp;code=xyz" data-source="[]" >
```
Please insert a linebreak before `data-source`

lutangar commented:
@core23 @greg0ire alright all fixed.

greg0ire commented:
Thanks @lutangar
